src/core/lib/iomgr/executor.cc

```diff
@@ -264,6 +264,10 @@ void Executor::ThreadMain(void* arg) {
     grpc_core::ExecCtx::Get()->InvalidateNow();
     subtract_depth = RunClosures(ts->name, closures);
   }
+#if !defined(__APPLE__)  // because of Thread local issues with MacOS and iOS
+  // See PR#19978
+  gpr_tls_set(&g_this_thread_state, reinterpret_cast<intptr_t>(nullptr));
+#endif  // !__APPLE__
```

Comment:
But i guess my question then is how will this code work for MacOS (and iOS C++) if those platforms actually do shutdown but don't properly support thread-local? Won't they just run into the failure that existed before this code was added? I guess it's ok for since we've  been bitten by this issue on Linux platforms but I just expect that this will create a latent bug that will arise at some time. That said, I guess it just maintains the status quo in that case so it does no harm.

Reply:
Just for my understanding, this PR does not fix thread-local issues on Mac and iOS, right? i.e. internal bug 138684610 is still unresolved?

Reply:
rmstar@, Yes this PR cleans up a thread local that we missed during shutdown but since this code fails with EINVAL with MacOS we have added the if !defined to avoid that issue

Reply:
vjpai@ It is as you are saying. It fixes a bug in our shutdown code for other platforms and maintains status quo for Apple platforms. As discussed offline, I do have a PR that would resolve this using C++ thread_local but that is not supported in our current version of XCode and we would need to move to Xcode9.2. More info in this PR #20413. I will add a todo to  apply that PR once we move to a later version of Xcode